bitbox-bridge/release/darwin/entitlements.plist

```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+    <!-- Sandbox app -->
+    <key>com.apple.security.app-sandbox</key>
```

Comment:
By trial an error, adding one key at a time, it turned out app-sandbox prevents the binary from executing after code signing:

```
$ lipo -create -output bitbox-bridge-uni bitbox-bridge-x86 bitbox-bridge-aarch64
$ file bitbox-bridge-uni 
bitbox-bridge-uni: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64]
bitbox-bridge-uni (for architecture x86_64):    Mach-O 64-bit executable x86_64
bitbox-bridge-uni (for architecture arm64):    Mach-O 64-bit executable arm64
```

Without the app-sandbox/true key, it runs ok:

```
$ codesign -f --timestamp --strict --entitlements entitlements.plist -s ... bitbox-bridge-uni
bitbox-bridge-uni: replacing existing signature
$ ./bitbox-bridge-uni --help
Set RUST_LOG=<filter> to enable logging. Example RUST_LOG=debug
BitBoxBridge 1.3.0
...
```

WITH the app sandbox enabled:

```
$ codesign -f --timestamp --strict --entitlements entitlements.plist -s ... bitbox-bridge-uni
bitbox-bridge-uni: replacing existing signature
$ ./bitbox-bridge-uni --help
Trace/BPT trap: 5
```

Any ideas why would this be an issue, due to libusb maybe?

Reply:
Maybe, I'm not sure, we use [hidapi](https://github.com/libusb/hidapi) which uses `IOHidManager` under the hood on osx. So I would assume that the usb entitlement gave us all permissions we needed.

After some googling it seems to me that the error message can have multiple possible root causes.

Is it possible to run it in lldb or something and get a backtrace?

hidapi is statically linked into the binary so on mac it only loads [apple libraries](https://github.com/ruabmbua/hidapi-rs/blob/master/build.rs#L123-L131).

Reply:
I also tried adding `com.apple.security.cs.allow-dyld-environment-variables` which I found in https://github.com/digitalbitbox/bitbox-wallet-app/blob/bdb9c97/frontends/qt/resources/MacOS/entitlements.plist but same result.

Reply:
Here's a crash report: https://gist.github.com/x1ddos/2571c83e87a195f927406cd95e9675c9

The error reads:

> Unable to get bundle identifier for container id bitbox-bridge-uni: Unable to get bundle identifier because code signature information has no Info.Plist.

It's unclear to me why this happens only with sandbox enabled.

Reply:
do we have to put an info.plist in the same dir as the executable? https://samwize.com/2019/10/13/info-plist-for-a-macos-command-line-app/

Reply:
hm... maybe but what would it contain, declare access to usb? I'm a bit confused though about the difference between entitlements.plist and an info.plist. I mean, the entitlements.plist already declares it. very strage.

Reply:
I don't know, I'm just guessing here. Maybe we can add something to this plist file? https://github.com/digitalbitbox/bitbox-bridge/blob/master/bitbox-bridge/release/darwin/ch.shiftcrypto.bitboxbridge.plist

Reply:
Tried notarizing without app-sandbox but got the error:

> The executable does not have the hardened runtime enabled.

Kind of expected, after all. Reading up on https://developer.apple.com/documentation/security/hardened_runtime....

Reply:
Ok, here's a working flow:

1. Remove app-sandbox from entitlements.plist.
2. Sign the universal bin:

```
codesign -f --timestamp --strict -o runtime \
  --entitlements entitlements.plist \
  -s <app-cert-identity> \
  bin/bitbox-bridge
```

3. Produce the installer with an extra `--sign` flag to productbuild:

```
productbuild --distribution ... --sign <installer-cert-identity> macos-installer.pkg
```

4. Submit the pkg for notarization:

```
xcrun altool --notarize-app \
  --primary-bundle-id ch.shiftcrypto.bitboxbridge \
  --username apple-dev@shiftcrypto.ch \
  --file macos-installer.pkg
```

5. Once notarized, staple the pkg:

```
xcrun stapler staple macos-installer.pkg
```

So, let's remove app-sandbox from entitlements.plist for now.